spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseExpireSnapshotsSparkAction.java

```diff
@@ -172,11 +179,14 @@ public BaseExpireSnapshotsSparkAction deleteWith(Consumer<String> newDeleteFunc)
 
       expireSnapshots.commit();
```

Comment:
The situation is like this:
expireSnapshots.commit() -> originalFiles.except(validFiles) (ongoing) -> _remove_orphan_files_ start and delete some file for building _originalFiles_ -> _expire_snapshots_ falis baceuse of FileNotFoundException.

Reply:
Intersting. 
I am rewording for my understanding...

`expireSnapshots.commit()` has removed snapshot entries from metadata. 
But the files (manifest list or manifest files) referring from that snapshot was present in `originalFiles` dataframe.
So, when `remove_orphan_files` called during that time, as metadata entry doesn't exist for those files, it has deleted it. 
So, when  `originalFiles.except(validFiles)` is called to actually compute the `originalFiles`, the manifest list or manifest files will be read which was deleted by `remove_orphan_files`. Leading to `FileNotFoundException`.

In your PR, `expireSnapshots` will remove the expired snapshot entires from the metadata but doesn't return list of removed files (which can potentially cause FileNotFoundException in above scenario)  and also doesn't delete the files as output list is empty. 
And these files will be later cleaned by running `remove_orphan_files` again. [Also note that cleaning expired files from `remove_orphan_files` is costlier (slower) compared to cleaning them in expiredSnapshots.

This PR can solve the issue. In my opinion, it looks hacky and user need to know when to set configuration to skipClean and when not to set it. I suggest we can block `remove_orphan_files` when `expireSnapshots` is running instead and vice versa (using event listener or some lock?). We can wait for others opinion.
cc: @rdblue , @RussellSpitzer

Reply:
or one more way to fix this is to just skip deleting the manifest file and manifest list but still delete the data files. But it is also hacky.

Reply:
I agree, we already have APIs for expiring without deleting and IMHO we should be pushing users towards deleting with expire snapshots. So anything that makes it seem like this is not the right way to clean old files should be avoided.

Reply:
Your understanding is quite right. But I don't think event listener or lock is a good idea, cause that seems need to introduce external dependencies.

Reply:
Agree. Your new solution of "trigger calculation for Dataset originalFiles before commit and cache it to reduce the probability of such failure" sounds good to me 👍🏻